prometheus/push.go

```diff
@@ -19,47 +19,153 @@
 
 package prometheus
 
-// Push triggers a metric collection by the default registry and pushes all
-// collected metrics to the Pushgateway specified by url. See the Pushgateway
-// documentation for detailed implications of the job and instance
-// parameter. instance can be left empty. You can use just host:port or ip:port
-// as url, in which case 'http://' is added automatically. You can also include
-// the schema in the URL. However, do not include the '/metrics/jobs/...' part.
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/common/model"
+)
+
+// Push works like PushFrom but collects the pushed metrics from the
+// DefaultRegistry.
+func Push(job string, grouping map[string]string, url string) error {
+	return PushFrom(DefaultRegistry, job, grouping, url)
+}
+
+// PushAdd works like PushAddFrom but collects the pushed metrics from the
+// DefaultRegistry.
+func PushAdd(job string, grouping map[string]string, url string) error {
+	return PushAddFrom(DefaultRegistry, job, grouping, url)
+}
+
+// PushFrom triggers a metric collection by the provided registry and pushes all
+// collected metrics to the Pushgateway specified by url, using the provided job
+// name and the (optional) further grouping labels (the grouping map may be
+// nil). See the Pushgateway documentation for detailed implications of the job
+// and other grouping labels. Neither the job name nor any grouping label value
+// may contain a "/". The metrics pushed must not contain a job label of their
+// own nor any of the grouping labels.
+//
+// You can use just host:port or ip:port as url, in which case 'http://' is
+//added automatically. You can also include the schema in the URL. However, do
+//not include the '/metrics/jobs/...' part.
 //
-// Note that all previously pushed metrics with the same job and instance will
-// be replaced with the metrics pushed by this call. (It uses HTTP method 'PUT'
-// to push to the Pushgateway.)
-func Push(job, instance, url string) error {
-	return defRegistry.Push(job, instance, url, "PUT")
+// Note that all previously pushed metrics with the same job and other grouping
+// labels will be replaced with the metrics pushed by this call. (It uses HTTP
+// method 'PUT' to push to the Pushgateway.)
+func PushFrom(r Registry, job string, grouping map[string]string, url string) error {
```

Comment:
That's a good point. As we are breaking the Push signature anyway, we can do that.

(It's in fact outright stupid to ever push the default registry as it contains all the process metrics.)

So done. (With the next push... no pun intended.)